LMS Step1 제출

cjsrhd94 opened the pull request:
안녕하세요 리뷰어님!

이번 미션을 진행하면서 궁금했던 점들이 있었습니다.
1) deleteHistory의 경우 특정 데이터를 삭제시 항상 호출된다고 생각하게된다고 봐도 무방할 것 같습니다. 그럼에도 개인적인 생각은 로그를 쌓는다는 행위가 핵심 비즈니스로직이 아니기 때문에 delete메서드 내에 묶지 않는 것이 옳다고 생각하였습니다. 리뷰어님 생각은 어떠신가요?

비록 짧은 시간이지만 잘부탁드립니다 :)

cjsrhd94 committed:
refactor: 질문 삭제 가능 검증 로직 리팩토링

cjsrhd94 committed:
refactor: 질문 삭제 메서드 로직 리팩토링

cjsrhd94 committed:
refactor: 대답 삭제 가능 검증 로직 리팩토링

cjsrhd94 committed:
chore: 미사용 메서드 삭제

cjsrhd94 committed:
refactor: 대답 삭제 메서드 로직 리팩토링

cjsrhd94 committed:
refactor: deleteQuestion 메서드 리팩토링

cjsrhd94 committed:
refactor: Answers 일급 컬렉션 추가 및 테스트 코드 작성

cjsrhd94 committed:
chore: 미사용 메서드 삭제

csh0034 commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -25,25 +25,9 @@ public class QnAService {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        Question question = questionRepository.findById(questionId)
```
메서드 리팩토링 잘 해주셨네요 👍

csh0034 commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,29 @@
+import nextstep.qna.CannotDeleteException;
+import nextstep.users.domain.NsUser;
+
+public class Answers {
```
일급컬렉션 적용 💯

csh0034 commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,29 @@
+	public List<DeleteHistory> deleteAnswers(NsUser loginUser, LocalDateTime now) throws CannotDeleteException {
+		List<DeleteHistory> deleteHistories = new ArrayList<>();
+		for (Answer answer : answers) {
+			answer.deleteAnswer(loginUser);
```
이전 미션에서 학습한 스트림을 적용하여 개선해보면 좋을것 같습니다.

csh0034 commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -35,54 +36,53 @@ public Question(Long id, NsUser writer, String title, String contents) {
-    public void addAnswer(Answer answer) {
-        answer.toQuestion(this);
-        answers.add(answer);
+    public List<DeleteHistory> deleteQuestion(NsUser loginUser, LocalDateTime now) throws CannotDeleteException {
```
테스트 관점에서 시간/랜덤 과 같은 대상은 테스트를 어렵게 하기 때문에 최대한
외부에서 주입되도록 개선하는것이 좋다고 생각합니다.

다만 이경우엔 필드명이 now 인점과 DeleteHistory 에 createdDate 는 필드에 생성시점에
값이 할당 되어있는걸로 보아 생성시에 외부에서 변경하지 못하도록 하기 위함으로
볼 수도 있는데요. 

now 라는 인자가 꼭필요한지에 대해 고민해보시면 좋을것 같습니다.

csh0034 commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -35,54 +36,53 @@ public Question(Long id, NsUser writer, String title, String contents) {
 
-    public boolean isOwner(NsUser loginUser) {
-        return writer.equals(loginUser);
+    private void delete(NsUser loginUser) throws CannotDeleteException {
```
CannotDeleteException 를 상위로 thorw 하지 않도록
`unchecked exception` 으로 적용해보면 좋을것 같습니다.

추가로 answers 에 deleteAnswers 와 동일하게 delete 처리를 할때
DeleteHistory 가 만들어지게 하면 좋을것 같아요.

createDeleteHistory 가 별도 메서드라면 실제 작성자 검증 + deleted 값이 변경되지
않았음에도 객체가 생성될수 있다고 보여집니다.

Answer 에 대해 DeleteHistory 가 생성되었다는건 `작성자 검증을 했으며 deleted 값이
변경되었다` 라고 보는것이 비지니스적으로 더 적합한것 같아요!

cjsrhd94 replied:
세번째 문단에 적어주신 말씀이 옳다고 생각합니다. 조언 감사합니다!

csh0034 commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -64,16 +60,27 @@ public NsUser getWriter() {
+        }
+    }
+
+    public DeleteHistory createDeleteHistory(LocalDateTime now) {
```
이부분도 Question 에 드린 코멘트와 같이 고민해보시면 좋을것 같습니다.

csh0034 reviewed (changes requested):
안녕하세요. 광현님
수강신청 미션을 함께 하게된 최승훈입니다. 

1단계 구현 관련하여 몇가지 코멘트 남겨두었는데 확인해주시고
다시 리뷰 요청 부탁드립니다!
질문에 대한 답변은 코멘트에 남겨두었습니다.

혹시 코멘트가 이해 안가시거나 추가로 궁금하신
사항 있으시면 편하게 DM 주세요 😄

cjsrhd94 committed:
fix: 피드백 반영

- CannotDeleteException UnchekcedException으로 변경
- createDate 관련 로직 수정
- delete시 DeleteHistory가 자동으로 생성되게 변경

csh0034 commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,25 @@
+
+	public List<DeleteHistory> deleteAnswers(NsUser loginUser) {
+		return answers.stream()
+			.map(a -> a.deleteAnswer(loginUser))
```
사소하지만 규칙에 맞춰 줄여쓰지 않는 연습을 해보면 좋을것 같아요!

- 규칙 5: 줄여쓰지 않는다(축약 금지).

csh0034 commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -35,54 +36,55 @@ public Question(Long id, NsUser writer, String title, String contents) {
-
-    public boolean isOwner(NsUser loginUser) {
-        return writer.equals(loginUser);
+        createDeleteHistory();
```
createDeleteHistory() 를 호출한후에 반환하게 해야되지 않을까요? 🤔

csh0034 commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -35,54 +36,55 @@ public Question(Long id, NsUser writer, String title, String contents) {
+        List<DeleteHistory> deleteHistories = new ArrayList<>();
+        // delete question
+        delete(loginUser);
+        deleteHistories.add(createDeleteHistory());
```
위의 코멘트와 같이 `delete(loginUser)` 에서 DeleteHistory 를 반환하게 하여
List 에 추가해보면 좋을것 같습니다.

csh0034 reviewed (approved):
리뷰 잘 반영해주셨네요 💯 
몇가지 코멘트남겨두었는데
다음단계 진행전에 적용해주시면 될 것 같아요!
2단계에서 뵙겠습니다 😄
